### PR TITLE
feat: add esbuild-why plugin

### LINKED ISSUE #302: Supporting `--why` for esbuild
Currently it does not:

```
 ERROR  Argument --why works only with @size-limit/webpack plugin and @size-limit/webpack-why plugin.
        You can add Bundle Analyzer to you own bundler.
```

Maybe it is possible now?

https://github.com/evanw/esbuild/issues/15
https://www.npmjs.com/package/esbuild-visualizer

Want to see how to make them work with `size-limit`

### PR DESCRIPTION
**Pull request by unional**
fixes #302

### CODE REVIEW

**Comment by unional** [comment]
Seems like the `check.esbuildMetafile` I added in `esbuild` didn't reach `esbuild-why` for some reason.

**Comment by ai** [comment]
I am on vacation and will review it a little later

File: packages/esbuild-why/index.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const { join } = require('path')
+const { writeFileSync } = require('fs')
+
+const self = {
+  name: '@size-limit/esbuild-why',
+
+  async step81(config, check) {
+    if (config.why && check.esbuildMetafile) {
+      if (!config.saveBundle) {
+        console.warn('`--why` option requires `--save-bundle` option')
```
Comment (ai) [inline packages/esbuild-why/index.js:11]: We need a test for this line

File: packages/esbuild-why/test/index.test.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    // let reportFile = join(config.checks[0].esbuildOutfile, 'report.html')
+    // let reportHTML = await readFile(reportFile)
+    // expect(reportHTML.toString()).toContain('superProject')
+  } finally {
+    await esbuild.finally(config, config.checks[0])
+  }
+})
+
+// it('applies both `modifyWebpackConfig`', async () => {
```
Comment (ai) [inline packages/esbuild-why/test/index.test.js:56]: Remove this test

File: packages/esbuild-why/test/index.test.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+//     let webpackConfig = config.checks[0].webpackConfig
+
+//     expect(webpackConfig.plugins).toContain(plugin)
+//   } finally {
+//     await webpack.finally(config, config.checks[0])
+//   }
+// })
+
+// it('supports --save-bundle', async () => {
```
Comment (ai) [inline packages/esbuild-why/test/index.test.js:88]: Remove this test

**Comment by unional** [comment]
Welcome back! I'll need to do some update to this PR, like updating the dependency as now the function is properly exposed, and other things.

btw, is the `step[81]` looks ok? I saw the code accepts 1-99 steps and there are some conventions, but I don't know where should this fit into (step 61 or 81 is my guess).

**Comment by ai** [comment]
> btw, is the step[81] looks ok?

There is no system. And since there is not a lot of plugins, just put it between necessary steps.

File: packages/esbuild-why/test/index.test.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    saveBundle: DIST,
+    checks: [{ files: [fixture('big.js')] }]
+  }
+  try {
+    await esbuild.before(config)
+    await esbuild.step20(config, config.checks[0])
+    await esbuild.step40(config, config.checks[0])
+    await esbuildWhy.step81(config, config.checks[0])
+
+    // let reportFile = join(config.checks[0].esbuildOutfile, 'report.html')
```
[the remark this answers is not in the document]
Comment (unional) [reply packages/esbuild-why/test/index.test.js:48]: The `esbuild-visualizer` does not add the project name to the resulting html, so need to see how to test this.

Also, `esbuild-visualizer` does not open the report automatically. Need to see how to handle that.

[the remark this answers is not in the document]
Comment (ai) [reply packages/esbuild-why/test/index.test.js:48]: Let’s check just for HTML file. For instance, by reading `<html>` or some word from the report

**Comment by ai** [comment]
Can we improve test coverage to fix CI?

**Comment by unional** [comment]
> Can we improve test coverage to fix CI?

Sure, but right now I'm trying to see how to use it locally. i.e. how to set the `--why` and `--saveBundle` so that I can see it actually working.

I'll try to do a `pnpm link` to one of my repos locally to test it out.

In the doc, it mention about the config for webpack but not the config for esbuild. And maybe that's also something to improve upon.

**Comment by ai** [comment]
> Sure, but right now I'm trying to see how to use it locally.

1. Install dependencies `pnpm install`
2. Create a test project in `fixtures/` with all plugins in `dependencies`
3. Call local `size-limit` script by `../package/size-limit/bin.js`

**Comment by unional** [comment]
> Install dependencies pnpm install
> Create a test project in fixtures/ with all plugins in dependencies
> Call local size-limit script by ../package/size-limit/bin.js

thx. I tried that but don't know if I'm doing it correctly.

I have added this in `/fixtures/esbuild-why/package.json`:

```js
{
  "name": "esbuild-why",
  "private": true,
  "devDependencies": {
    "@size-limit/esbuild-why": ">= 0.0.0",
    "@size-limit/file": ">= 0.0.0",
    "size-limit": ">= 0.0.0"
  },
  "size-limit": [
    {
      "path": "index.js"
    }
  ]
}
```
with an empty `index.js`.

> Call local size-limit script by ../package/size-limit/bin.js

I can't find where should I do this. From the path, seems like you are referring to calling `size-limit` from `/packages/esbuild-why`?

I'll find some time to work on this PR this week.

btw, I always get this timeout error when running the tests. I'm using WSL in Windows:

```sh
packages/time/test/get-running-time.test.js (15.731 s)
  ● calculates running time

    thrown: "Exceeded timeout of 15000 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

      15 | })
      16 |
    > 17 | it('calculates running time', async () => {
         | ^
      18 |   let runTime = await getRunningTime(EXAMPLE)
      19 |   expect(runTime).toBeGreaterThan(0.01)
      20 |   expect(runTime).toBeLessThan(0.5)

      at Object.it (packages/time/test/get-running-time.test.js:17:1)
```

**Comment by ai** [comment]
> I can't find where should I do this.

You should be in the fixture folder and call:

```sh
$ cd
size-limit/fixtures/integration
$ ../../packages/size-limit/bin.js
✔ Adding to empty webpack project
✔ Running JS in headless Chrome
  
  Package size is 58 B less than limit
  Size limit:   200 B
  Size:         142 B with all dependencies, minified and gzipped
  Loading time: 10 ms on slow 3G
  Running time: 84 ms on Snapdragon 410
  Total time:   94 ms
```

**Comment by ai** [comment]
> btw, I always get this timeout error when running the tests. I'm using WSL in Windows:

I don’t have a Windows to investigate the reason without your help. But let’s ignore this test until we will not finish this PR.

**Comment by unional** [comment]
> I don’t have a Windows to investigate the reason without your help. But let’s ignore this test until we will not finish this PR.

Sure, I'm ignoring that. :)

**Comment by unional** [comment]
Hi @ai, sorry for the long wait.

I have updated the code a bit to better handle the error message.
I couldn't add the test needed because I can't find the way to write it without some refactoring:

```js
// e.g.
it('throws on --why argument without webpack', async () => {
  expect(await error('file', ['--why'])).toMatchSnapshot()
})

// I need
it('throws on --why argument with esbuild but without esbuild-why', async () => {
  expect(await error(/* don't know how to do it here */, ['--why'])).toMatchSnapshot()
})
```

**Comment by ai** [comment]
Nice. I fixed CI. Will review PR tomorrow (need to run to a party)

**Comment by ai** [comment]
Thanks! Released in 8.2.

**Comment by unional** [comment]
🎉

**Pull request merged**